File: lessons/ch1_7.cpp
// Глава 1.7 · Обновление: $set, $inc, $unset — заготовка для примеров.
//
// Перед главой верните базы в исходное состояние (глава 1.0, §8).
// Пример из главы вставьте в фигурные скобки в конце main вместо проверочных строк
// и запустите:
//
//     cd lessons
//     g++ -std=c++17 ch1_7.cpp -o ch1_7 $(pkg-config --cflags --libs libmongocxx1) && ./ch1_7   # macOS, Homebrew
//     docker compose run --rm cpp lessons/ch1_7.cpp       # то же в контейнере
//
// Если пример — целая функция, его место — над main, у отметки.
// Следующий пример вставляйте вместо предыдущего: каждый пример запускается
// один раз и по порядку — именно так получены выводы в справочнике.

#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <optional>
#include <sstream>
#include <string>
#include <vector>

#include <bsoncxx/builder/basic/array.hpp>
#include <bsoncxx/builder/basic/document.hpp>
#include <bsoncxx/builder/basic/kvp.hpp>
#include <bsoncxx/json.hpp>
#include <bsoncxx/types.hpp>
#include <mongocxx/client.hpp>
#include <mongocxx/exception/bulk_write_exception.hpp>
#include <mongocxx/exception/operation_exception.hpp>
#include <mongocxx/instance.hpp>
#include <mongocxx/options/find.hpp>
#include <mongocxx/options/insert.hpp>
#include <mongocxx/options/update.hpp>
#include <mongocxx/uri.hpp>

using bsoncxx::builder::basic::kvp;
using bsoncxx::builder::basic::make_array;
using bsoncxx::builder::basic::make_document;

// ── функции из примеров главы ─────────────────────────────────────


int main() {
    mongocxx::instance instance{};
    const char* env = std::getenv("MONGO_URI");
    mongocxx::client client{mongocxx::uri{env ? env : "mongodb://localhost:27017"}};

    auto db = client["shop"];
    auto products = db["products"];
    auto orders = db["orders"];
    auto box = client["sandbox"]["products"];   // песочница: здесь можно менять

    // Товары, которые добавлялись в песочницу в главе 1.3. Если песочницу
    // сбрасывали, заготовка добавит их снова — обновлять будет что.
    if (box.count_documents(make_document(kvp("_id", "p-101"))) == 0) {
        std::vector<bsoncxx::document::value> added;
        added.push_back(make_document(kvp("_id", "p-101"), kvp("title", "Подставка для ноутбука"),
                                      kvp("category", "аксессуары"), kvp("price", 2490)));
        added.push_back(make_document(kvp("_id", "p-102"), kvp("title", "Чехол для планшета"),
                                      kvp("category", "аксессуары"), kvp("price", 1890)));
        added.push_back(make_document(kvp("sku", "SKU-AC-022"), kvp("title", "Коврик для мыши XL"),
                                      kvp("brand", "OEM"), kvp("category", "аксессуары"), kvp("price", 1290)));
        box.insert_many(added);
    }

    auto orders_box = client["sandbox"]["orders"];   // копия заказов: их можно менять
    if (orders_box.count_documents(make_document()) == 0) {
        std::vector<bsoncxx::document::value> all;
        for (auto&& doc : orders.find(make_document())) all.emplace_back(doc);
        orders_box.insert_many(all);
    }

    auto stats = client["sandbox"]["product_stats"];   // счётчики просмотров для §6
    stats.drop();                                      // каждый запуск заготовки считает с нуля
    auto now = std::time(nullptr);
    std::ostringstream day;
    day << std::put_time(std::gmtime(&now), "%Y-%m-%d");
    std::string today = day.str();

    {
        // ── пример из главы ──────────────────────────────────────
        // Проверочные строки: замените их примером из главы.
        client["admin"].run_command(make_document(kvp("ping", 1)));
        std::cout << "Заготовка главы 1.7 подключилась к серверу. Замените проверочные строки примером из главы." << std::endl;
    }
}

File: lessons/ch1_7.go
// Глава 1.7 · Обновление: $set, $inc, $unset — заготовка для примеров.
//
// Перед главой верните базы в исходное состояние (глава 1.0, §8).
// Пример из главы вставьте в фигурные скобки в конце main вместо проверочных строк
// и запустите:
//
//	cd lessons
//	go run ch1_7.go
//
// Если пример начинается с func или type, его место — над main, у отметки.
// Следующий пример вставляйте вместо предыдущего: каждый пример запускается
// один раз и по порядку — именно так получены выводы в справочнике.
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
)

// Импорты нужны разным примерам главы. Строки ниже не дают Go ругаться
// на те, которыми текущий пример не пользуется.
var _ = fmt.Sprint
var _ = time.Now
var _ = bson.D{}
var _ = mongo.ErrNoDocuments

// ── функции и типы из примеров главы ─────────────────────────────

// Product — поля товара, которые читают примеры. Тип объявлен в главе 1.4.
type Product struct {
	ID       string `bson:"_id"`
	SKU      string `bson:"sku"`
	Title    string `bson:"title"`
	Category string `bson:"category"`
	Price    int    `bson:"price"`
}


func main() {
	uri := os.Getenv("MONGO_URI")
	if uri == "" {
		uri = "mongodb://localhost:27017"
	}
	client, err := mongo.Connect(options.Client().ApplyURI(uri))
	if err != nil {
		log.Fatal(err)
	}
	ctx := context.Background()
	defer client.Disconnect(ctx)

	db := client.Database("shop")
	products := db.Collection("products")
	orders := db.Collection("orders")
	box := client.Database("sandbox").Collection("products") // песочница: здесь можно менять
	_, _, _, _ = db, products, orders, box

	// Товары, которые добавлялись в песочницу в главе 1.3. Если песочницу
	// сбрасывали, заготовка добавит их снова — обновлять будет что.
	if n, _ := box.CountDocuments(ctx, bson.D{{Key: "_id", Value: "p-101"}}); n == 0 {
		box.InsertMany(ctx, []any{
			bson.D{{Key: "_id", Value: "p-101"}, {Key: "title", Value: "Подставка для ноутбука"}, {Key: "category", Value: "аксессуары"}, {Key: "price", Value: 2490}},
			bson.D{{Key: "_id", Value: "p-102"}, {Key: "title", Value: "Чехол для планшета"}, {Key: "category", Value: "аксессуары"}, {Key: "price", Value: 1890}},
			bson.D{{Key: "sku", Value: "SKU-AC-022"}, {Key: "title", Value: "Коврик для мыши XL"}, {Key: "brand", Value: "OEM"}, {Key: "category", Value: "аксессуары"}, {Key: "price", Value: 1290}},
		})
	}

	ordersBox := client.Database("sandbox").Collection("orders") // копия заказов: их можно менять
	if n, _ := ordersBox.CountDocuments(ctx, bson.D{}); n == 0 {
		var all []bson.D
		cursor, _ := orders.Find(ctx, bson.D{})
		cursor.All(ctx, &all)
		docs := make([]any, len(all))
		for i := range all {
			docs[i] = all[i]
		}
		ordersBox.InsertMany(ctx, docs)
	}

	stats := client.Database("sandbox").Collection("product_stats") // счётчики просмотров для §6
	stats.Drop(ctx)                                                 // каждый запуск заготовки считает с нуля
	today := time.Now().Format("2006-01-02")
	_, _, _ = ordersBox, stats, today

	{
		// ── пример из главы ──────────────────────────────────────
		// Проверочные строки: замените их примером из главы.
		if err := client.Ping(ctx, nil); err != nil {
			log.Fatal(err)
		}
		fmt.Println("Заготовка главы 1.7 подключилась к серверу. Замените проверочные строки примером из главы.")
	}
}
